### PR TITLE
Test for nonce reuse

### PR DESCRIPTION
**Pull request by nickfarrow**
Resolves https://github.com/frostsnap/frost-esp32/issues/14
Builds on and based on testing approach in https://github.com/frostsnap/frost-esp32/pull/66

### CODE REVIEW

**Comment by LLFourn** [comment]
Hey can we put all the tests in `malicious.rs` that are to do with a malicious device or coordinator. What's the difference between this new malicious poly test and the old one?

File: frostsnap_core/tests/malicious.rs [hunk 1]
```diff
@@ -34,3 +36,56 @@ fn keygen_maliciously_replace_public_poly() {
...
+        .unwrap();
+
+    let _device_responses = device
+        .recv_coordinator_message(sign_request.clone())
+        .unwrap();
+
+    let _device_sends = device.sign_ack(true).unwrap();
+
+    // Receive the same sign request again with the same nonce
+    let sign_request_result = device.recv_coordinator_message(sign_request.clone());
```
Comment (LLFourn) [inline frostsnap_core/tests/malicious.rs:80]: Nit: shouldn't this be a different sign request. If it's precisely the same one it's technically fine to carry on right?. I guess in practice we will not "recall" the sign request that occurs so even the same sign request will throw an error so maybe to fine to leave as is.

File: frostsnap_core/tests/malicious.rs [hunk 2]
```diff
@@ -34,3 +36,56 @@ fn keygen_maliciously_replace_public_poly() {
...
+
+    // TODO: Specific error messages and match on NonceReuse
```
Comment (LLFourn) [inline frostsnap_core/tests/malicious.rs:82]: Not sure this is a good idea see: https://github.com/frostsnap/frostsnap/pull/91#issuecomment-1732983696

**Review by LLFourn — APPROVED** [review]
Thanks! Left two minor comments.

**Comment by nickfarrow** [comment]
Now creates a new sign request and that has the same nonces as the previous signing session

File: frostsnap_core/tests/malicious.rs [hunk 3]
```diff
@@ -34,3 +36,69 @@ fn keygen_maliciously_replace_public_poly() {
...
+        .start_sign(SignTask::Plain(b"utxo.club!".to_vec()), device_ids.clone())
+        .unwrap();
+
+    let _device_responses = device
+        .recv_coordinator_message(sign_request.clone())
+        .unwrap();
+
+    let _device_sends = device.sign_ack(true).unwrap();
+
+    // Receive the same sign request again with the same nonces
```
Comment (LLFourn) [inline frostsnap_core/tests/malicious.rs:79]:
```suggestion
    // Receive the same sign request again with the same nonces
```
This comment is outdated :)

**Pull request merged**